File: lib/presentation/widgets/circular_percent.dart
import 'package:flutter/material.dart';

/// An animated circular progress ring with a centered percentage label.
class CircularPercent extends StatelessWidget {
  const CircularPercent({
    super.key,
    required this.value,
    this.size = 96,
    this.strokeWidth = 9,
    this.color,
    this.label,
  }) : assert(value >= 0 && value <= 1);

  final double value;
  final double size;
  final double strokeWidth;
  final Color? color;
  final String? label;

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final accent = color ?? scheme.primary;
    return SizedBox(
      width: size,
      height: size,
      child: TweenAnimationBuilder<double>(
        tween: Tween(begin: 0, end: value),
        duration: const Duration(milliseconds: 700),
        curve: Curves.easeOutCubic,
        builder: (context, v, _) => Stack(
          alignment: Alignment.center,
          children: [
            SizedBox.expand(
              child: CircularProgressIndicator(
                value: v,
                strokeWidth: strokeWidth,
                backgroundColor: scheme.surfaceContainerHighest,
                valueColor: AlwaysStoppedAnimation(accent),
                strokeCap: StrokeCap.round,
              ),
            ),
            Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(
                  '${(v * 100).round()}%',
                  style: TextStyle(
                    fontSize: size * 0.22,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                if (label != null)
                  Text(
                    label!,
                    style: TextStyle(
                      fontSize: size * 0.12,
                      color: scheme.onSurfaceVariant,
                    ),
                  ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

File: android_native/app/src/main/java/com/example/widgetgallery/presentation/components/CircularPercent.kt
package com.example.widgetgallery.presentation.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.roundToInt

/** Animated circular progress ring with a centered percentage label. */
@Composable
fun CircularPercent(
    value: Float,
    modifier: Modifier = Modifier,
    size: Dp = 96.dp,
    label: String? = null,
) {
    val animated by animateFloatAsState(
        targetValue = value.coerceIn(0f, 1f),
        animationSpec = tween(700), label = "ring",
    )
    Box(modifier.size(size), contentAlignment = Alignment.Center) {
        CircularProgressIndicator(
            progress = { animated },
            modifier = Modifier.fillMaxSize(),
            strokeWidth = 9.dp,
            trackColor = MaterialTheme.colorScheme.surfaceContainerHighest,
        )
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text("${(animated * 100).roundToInt()}%",
                fontSize = (size.value * 0.22f).sp,
                fontWeight = FontWeight.Bold)
            if (label != null) {
                Text(label, fontSize = (size.value * 0.12f).sp,
                    color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
        }
    }
}
